### ISSUE TITLE
no-extra-semi does not fix multiple semicolons

### ISSUE THREAD

**Issue by gajus** [eslint/eslint#5004]
[`no-extra-semi`](http://eslint.org/docs/rules/no-extra-semi) fix only removes one extra semicolon.

eslint@1.10.3

Input:

```js
var test;;;;;
```

Output:

```js
var test;;;;
```

Expected output:

```js
var test;
```

**Comment by eslintbot** [comment]
@gajus Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
Its because of the overlapping range information for the fixes. This change was introduced here: https://github.com/eslint/eslint/pull/4329

**Comment by gajus** [comment]
@gyandeeps Is this available in 2.0.0 beta branch? (or anywhere else for that matter)

**Comment by gyandeeps** [comment]
The above change was introduced in 1.9 and that what we have now in current master. Because it was not safe to make fixes which had overlapping range info based on ast.

**Comment by gajus** [comment]
I am running `eslint@1.10.3`. Then shouldn't this be fixed?

**Comment by gyandeeps** [comment]
I think we are talking in 2 different directions. What you are seeing right now (based on the issue), its behaving as expected (as per the code for fixer).

Before 1.9, we used to apply fixes which used to have overlapping range info (based on ast) also. But later on we found out that it was not safe all the time (ref issue: https://github.com/eslint/eslint/issues/4321). That when we made a change to our fixer code to not apply fixes which have overlapping range info.

With that, the fixes for the code above has overlapping range, that why it not applying all the fixes.

**Comment by gajus** [comment]
"overlapping range info" is not very descriptive.

If input is `var test;;;;;` and rule is supposed to remove multiple (redundant) semicolons, then expected output is `var test;`. If output is anything else (`var test;;;;` in this case) then it is a bug.

Furthermore, iterating the ESLint `--fix` multiple times produces different result (each time removing an additional column in this case), which is not a predictable behaviour.

This does look like a bug.

**Comment by gyandeeps** [comment]
Here is info for overlapping range:

```json
{
      "type": "Punctuator",
      "value": ";",
      "range": [
        8,
        9
      ]
    },
    {
      "type": "Punctuator",
      "value": ";",
      "range": [
        9,
        10
      ]
    },
    {
      "type": "Punctuator",
      "value": ";",
      "range": [
        10,
        11
      ]
    },
    {
      "type": "Punctuator",
      "value": ";",
      "range": [
        11,
        12
      ]
    },
    {
      "type": "Punctuator",
      "value": ";",
      "range": [
        12,
        13
      ]
    }
```

**Comment by michaelficarra** [comment]
@gyandeeps That is not overlapping. One end of the range is inclusive, and the other is exclusive

**Comment by nzakas** [comment]
@michaelficarra based on our current implementation, these are overlapping because we had some trouble with ranges that touch like this. We made a fix to prevent invalid code from being created and are still trying to find a better implementation.

**Comment by gajus** [comment]
A bit naive suggestion, but what about simply running the fix logic multiple times until next iteration output is not different from its input?

**Comment by platinumazure** [comment]
@michaelficarra I took a crack at trying to address the issue of adjacent ranges differently, but it introduced other problems and had to be reverted. My thought at the time was that the adjacent ranges that had problems actually was a symptom of the ESLint rules not being optimally designed and relied on fixes being applied in a different order than what the fix logic does. I'd love to take another look at this but I don't know when I'll have time.

**Comment by gajus** [comment]
Just in case, this is what I have meant:

```js
import _ from 'lodash';

import {
    CLIEngine
} from 'eslint';

let cli,
    fixText;

cli = new CLIEngine({
    allowInlineConfig: true,
    baseConfig: false,
    configFile: require.resolve('./eslintrc.json'),
    envs: [],
    extensions: [],
    fix: true,
    globals: [],
    parser: require.resolve('babel-eslint'),
    rulePaths: [],
    useEslintrc: false
});

fixText = (text: string): string => {
    let fatalError,
        report;

    report = cli.executeOnText(text).results[0];

    fatalError = _.find(report.messages, {
        fatal: true
    });

    if (fatalError) {
        throw new Error(fatalError.message + ' (line: ' + fatalError.line + ', column: ' + fatalError.column + ').');
    }

    return report.output || text;
};

export default (text: string): string => {
    let input,
        maxIterationCount,
        output;

    maxIterationCount = 100;
    output = text;

    // @see https://github.com/eslint/eslint/issues/5004

    do {
        input = output;

        output = fixText(input);

        if (--maxIterationCount < 0) {
            throw new Error('Max iteration count.');
        }
    } while (output !== input);

    return output;
};

```

**Comment by alberto** [comment]
@gajus the reason for not doing that is performance

**Comment by gajus** [comment]
Performance always comes second to functionality.

**Comment by nzakas** [comment]
@gajus not really. People won't use something that's slow even if they like the result. It's a balance between the two that we have to be careful of.

**Comment by gajus** [comment]
Lets try a real world analogy: No one would drive a car that is known to malfunction over a slower but safe car.

**Comment by nzakas** [comment]
If it's faster to walk than take the car, people will walk. :)

We've explored running several passes before, but the perf impact coupled with the potential for creating new problems makes it infeasible right now. I'll take another look after 2.0.0. In the meantime, people who want to run it multiple times can already choose to do so manually.

**Comment by ilyavolodin** [comment]
This has been proposed as one of the points to focus on in 3.0.0 timeframe. I'll close this issue for now, since there isn't much we can do to fix it before we address larger issues with autofix.
